File: src/services/mpesaService.js
const axios = require('axios');
const db = require('../config/database');

class MpesaService {
  constructor() {
    this.baseUrl = process.env.MPESA_ENVIRONMENT === 'production' 
      ? 'https://api.safaricom.co.ke'
      : 'https://sandbox.safaricom.co.ke';
  }

  async getAccessToken() {
    const auth = Buffer.from(
      `${process.env.MPESA_CONSUMER_KEY}:${process.env.MPESA_CONSUMER_SECRET}`
    ).toString('base64');

    const response = await axios.get(
      `${this.baseUrl}/oauth/v1/generate?grant_type=client_credentials`,
      { headers: { Authorization: `Basic ${auth}` } }
    );

    return response.data.access_token;
  }

  async registerUrls() {
    try {
      const token = await this.getAccessToken();

      const response = await axios.post(
        `${this.baseUrl}/mpesa/c2b/v1/registerurl`,
        {
          ShortCode: process.env.MPESA_SHORTCODE,
          ResponseType: 'Completed',
          ConfirmationURL: process.env.MPESA_CONFIRMATION_URL,
          ValidationURL: process.env.MPESA_VALIDATION_URL,
        },
        { headers: { Authorization: `Bearer ${token}` } }
      );

      return response.data;
    } catch (error) {
      console.error('M-PESA URL registration failed:', error.response?.data || error.message);
      throw error;
    }
  }

  async simulatePayment(phoneNumber, amount, accountNumber) {
    try {
      const token = await this.getAccessToken();

      const response = await axios.post(
        `${this.baseUrl}/mpesa/c2b/v1/simulate`,
        {
          ShortCode: process.env.MPESA_SHORTCODE,
          CommandID: 'CustomerPayBillOnline',
          Amount: amount,
          Msisdn: phoneNumber,
          BillRefNumber: accountNumber,
        },
        { headers: { Authorization: `Bearer ${token}` } }
      );

      return response.data;
    } catch (error) {
      console.error('M-PESA simulation failed:', error.response?.data || error.message);
      throw error;
    }
  }
}

module.exports = new MpesaService();

File: app/services/mpesa_service.py
"""
SMART BIN LTD - M-PESA SERVICE (Safaricom Daraja API)
"""
import base64
import json
from datetime import datetime
from typing import Optional
import requests
from app.config import get_settings

settings = get_settings()

class MpesaService:
    def __init__(self):
        self.consumer_key = settings.mpesa_consumer_key
        self.consumer_secret = settings.mpesa_consumer_secret
        self.shortcode = settings.mpesa_shortcode
        self.passkey = settings.mpesa_passkey
        self.environment = settings.mpesa_environment
        self.base_url = (
            "https://sandbox.safaricom.co.ke"
            if self.environment == "sandbox"
            else "https://api.safaricom.co.ke"
        )
        self.mock_mode = not self.consumer_key or self.consumer_key == "your-consumer-key"

        if self.mock_mode:
            print("[M-PESA] MOCK MODE: No credentials. Operations log to console only.")

    def get_access_token(self) -> Optional[str]:
        if self.mock_mode:
            return "mock-token"
        url = f"{self.base_url}/oauth/v1/generate?grant_type=client_credentials"
        try:
            resp = requests.get(url, auth=(self.consumer_key, self.consumer_secret), timeout=10)
            return resp.json().get("access_token")
        except Exception as e:
            print(f"[M-PESA] Token error: {e}")
            return None

    def register_urls(self, validation_url: str, confirmation_url: str) -> dict:
        if self.mock_mode:
            print(f"[M-PESA MOCK] Would register URLs:")
            print(f"  Validation: {validation_url}")
            print(f"  Confirmation: {confirmation_url}")
            return {"status": "mock-registered"}

        token = self.get_access_token()
        if not token:
            return {"error": "Could not get access token"}

        url = f"{self.base_url}/mpesa/c2b/v1/registerurl"
        headers = {"Authorization": f"Bearer {token}"}
        payload = {
            "ShortCode": self.shortcode,
            "ResponseType": "Completed",
            "ConfirmationURL": confirmation_url,
            "ValidationURL": validation_url
        }
        resp = requests.post(url, json=payload, headers=headers, timeout=10)
        return resp.json()

    def validate_payment(self, account_number: str, amount: str, phone: str) -> dict:
        if self.mock_mode:
            print(f"[M-PESA MOCK] Validating: Acc={account_number}, Amount={amount}, Phone={phone}")
        return {"ResultCode": 0, "ResultDesc": "Accepted"}

    def confirm_payment(self, data: dict) -> dict:
        if self.mock_mode:
            print(f"[M-PESA MOCK] Confirmation received:")
            print(json.dumps(data, indent=2))
        return {"ResultCode": 0, "ResultDesc": "Success"}

mpesa_service = MpesaService()
